test/controllers/CancelLeaveSchemeControllerSpec.scala

```diff
@@ -0,0 +1,201 @@
+
+      val application = applicationBuilder(
+        userAnswers = Some(emptyUserAnswers),
+        registration = registrationNoLongerSuppliesExclusion
+      ).overrides(bind[RegistrationService].toInstance(mockRegistrationService)).build()
+
+      when(mockRegistrationService.amendRegistration(any(), any(), any(), any())(any())) thenReturn Future.successful(Right(()))
+
+      running(application) {
+        val request = FakeRequest(POST, cancelLeaveSchemeRoute).withFormUrlEncodedBody(("value", "true"))
+
+        val result = route(application, request).value
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual CancelLeaveSchemeCompletePage.route(emptyWaypoints).url
+      }
+    }
+
+    "must redirect to /your-account when the user submits false and is not cancelling their request to leave the scheme" in {
+      val application = applicationBuilder(
+        userAnswers = Some(emptyUserAnswers),
+        registration = registrationNoLongerSuppliesExclusion
+      ).build()
+
+      running(application) {
+        val request = FakeRequest(POST, cancelLeaveSchemeRoute).withFormUrlEncodedBody(("value", "false"))
+
+        val config = application.injector.instanceOf[FrontendAppConfig]
+
+        val result = route(application, request).value
+
+        status(result) mustEqual SEE_OTHER
+        redirectLocation(result).value mustEqual config.iossYourAccountUrl
+      }
+    }
+
+    "must return a Bad Request and errors when invalid data is submitted" in {
+
+      val application = applicationBuilder(
+        userAnswers = Some(emptyUserAnswers),
+        registration = registrationNoLongerSuppliesExclusion
+      ).build()
+
+      running(application) {
+        val request = FakeRequest(POST, cancelLeaveSchemeRoute).withFormUrlEncodedBody(("value", ""))
+
+        val boundForm = form.bind(Map("value" -> ""))
+
+        val view = application.injector.instanceOf[CancelLeaveSchemeView]
+
+        val result = route(application, request).value
+
+        status(result) mustEqual BAD_REQUEST
+        contentAsString(result) mustEqual view(boundForm, emptyWaypoints)(request, messages(application)).toString
+      }
+    }
+
+
+    "must return a Bad Request and errors when a trader tries to reverse their exclusion and the effective date has passed" in {
+
+      val noLongerSuppliesExclusion = EtmpExclusion(
+        NoLongerSupplies,
+        LocalDate.now(stubClock),
+        LocalDate.now(stubClock).minusDays(1),
+        false
+      )
+
+      val application = applicationBuilder(
+        userAnswers = Some(emptyUserAnswers),
+        registration = registrationWrapper.copy(registration = registrationWrapper.registration.copy(exclusions = Seq(noLongerSuppliesExclusion)))
+      ).build()
+
+      running(application) {
+
+        val request = FakeRequest(GET, cancelLeaveSchemeRoute)
+
+        val result = route(application, request).value
+
+        status(result) mustEqual SEE_OTHER
```

Comment:
Shouldn't this be returning BadRequest?

Reply:
Good catch, it should return a SEE_OTHER / Redirect, but the test description was wrong. Fixed now